src/StatsTools/variational_energy_estimator.jl

```diff
@@ -0,0 +1,76 @@
+"""
+    variational_energy_estimator(shifts, overlaps; skip=0, kwargs...)
+    variational_energy_estimator(df::DataFrame; max_replicas=:all, skip=0, kwargs...)
+    -> r::RatioBlockingResult
+Compute the variational energy estimator from the replica time series of the `shifts` and
+coefficient vector `overlaps` by blocking analysis.
+The keyword argument `max_replicas` can be used to constrain the number of replicas
+processed to be smaller than all available in `df`.
+Other keyword arguments are passed on to [`ratio_of_means()`](@ref).
+Returns a [`RatioBlockingResult`](@ref).
+
+An estimator for the variational energy
+```math
+\\frac{⟨\\mathbf{c}⟩^† \\mathbf{H}⟨\\mathbf{c}⟩}{⟨\\mathbf{c}⟩^†⟨\\mathbf{c}⟩}
+```
+is calculated from
+```math
+Ē_{v}  =  \\frac{\\sum_{a<b}^R \\overline{(S_a+S_b) \\mathbf{c}_a^† \\mathbf{c}_b}}
+               {2\\sum_{a<b}^R \\overline{\\mathbf{c}_a^† \\mathbf{c}_b}} ,
+```
+where the sum goes over distinct pairs out of the ``R`` replicas. See
+[arXiv:2103.07800](http://arxiv.org/abs/2103.07800).
+
+The `DataFrame` version can extract the relevant information from the result of
+[`lomc!`](@ref). Set up [`lomc!`](@ref) with the keyword argument
+`replica = AllOverlaps(R)` and `R ≥ 2`.
+
+See [`AllOverlaps`](@ref).
+"""
+function variational_energy_estimator(shifts, overlaps; kwargs...)
+    num_replicas = length(shifts)
+    if length(overlaps) ≠ binomial(num_replicas, 2)
+        throw(ArgumentError(
+            "The number of `overlaps` needs to be `binomial(length(shifts),2)`."
+        ))
+    end
+    denominator = sum(overlaps)
+    numerator = zero(shifts[1])
+    count_overlaps = 0
+    for i in 1:num_replicas, j in i+1:num_replicas
+        count_overlaps += 1
+        @. numerator += 1 / 2 * (shifts[i] + shifts[j]) * overlaps[count_overlaps]
+    end
```

Comment:
This has the potential for undefined behaviour. You have a check for the length of `overlaps` but this part also assumes that `overlaps` is ordered in a certain way. If the DataFrame method of `variational_energy_estimator` is used then it is OK, but if this method is used as is then it requires the user to build `overlaps` correctly.

At the very least the doc string should have a warning and a suggestion to use the DataFrame method. Alternatively, another method could take `overlaps` as an array and unwrap the replicas in the correct order and then this loop doesn't need to introduce an incrementing counter on top of the loop index.

Reply:
I have added a comment in the docstring as suggested. 

I agree that the loop is not elegant, but I couldn't think of something better when I wrote it. And it works. Otherwise you would need to have a matrix of overlaps. In order to avoid storing redundant data, you would need a data structure that represents a triangular matrix (without the diagonal), but this all seemed a bit over-the-top for what is needed here. Alternatively, we could only export/document the `DataFrame` method.

Reply:
I have also re-written the reference to `AllOverlaps` to use a keyword argument. This means that we should merge PR #152 asap.